# Лабораторная работа 1. Базовый синтаксис языка Python 3.

# Справочная информация:

# 1. Типы данных

Основными типами данных в любом языке программирования являются:

целые числа, вещественные числа, строки (символы), логические переменные (True (истина) или False (ложь)).

Они могут быть записаны следующим образом

In [2]:
1, 1.0, '1', 1<2

(1, 1.0, '1', True)

Функцией type возвращает тип аргумента: 

In [3]:
type(1), type(1.0), type('1'), type(1<2)

(int, float, str, bool)

# 1.2 Операции с числами: $A+B$, $A-B$, $A*B$, $A/B$, $A^B$

In [3]:
# комментарий: с целыми числами
A=1 
B=2
print (' A + B = ',A+B) # сумма
print (' A - B = ',A-B) # разность;
print (' A * B = ',A*B) # произведение;
print (' A / B = ',A/B) # частное;
print (' A ^ B = ',A**B) # возведение в степень.

 A + B =  3
 A - B =  -1
 A * B =  2
 A / B =  0.5
 A ^ B =  1


In [4]:
# комментарий: с вещественными числами
A=1.0 
B=2.0
print (' A + B  = ',A+B)  # сумма
print (' A - B  = ',A-B)  # разность;
print (' A * B  = ',A*B)  # произведение;
print (' A / B  = ',A/B)  # частное;
print (' A % B  = ',A%B)  #  остаток от деления;
print (' A // B = ',A//B) # деление без остатка;
print (' A ^ B  = ',A**B) # возведение в степень.

 A + B  =  3.0
 A - B  =  -1.0
 A * B  =  2.0
 A / B  =  0.5
 A % B  =  1.0
 A // B =  0.0
 A ^ B  =  1.0


In [4]:
# комментарий: ещё о целых числах, без автоматического переопределения типа было бы так
A=1
B=2
print (' A / B = ',A/B, 'возвращаемый тип - ', type(A/B)) # автоматическое переопределение;
print (' A / B = ',(int)(A/B), 'возвращаемый тип - ', type((int)(A/B))) # без автоматического переопределения;

 A / B =  0.5 возвращаемый тип -  <class 'float'>
 A / B =  0 возвращаемый тип -  <class 'int'>


# 1.2 Операции над строками

In [5]:
Ast=' я люблю ' 
Bst='информатику '
n=3
print (' Ast + Bst = ',Ast+Bst)  # конкатенация 
print ('(Ast + Bst) * n = ',(Ast+Bst)*n)  # повторение n раз

 Ast + Bst =   я люблю информатику 
(Ast + Bst) * n =   я люблю информатику  я люблю информатику  я люблю информатику 


# 1.3 Логические операции

In [2]:
A=1.0 
B=2.0
print (' A >  B  ? ',A>B)   # больше
print (' A <  B  ? ',A<B)   # меньша
print (' A >= B  ? ',A>=B)  # больше или равно
print (' A <= B  ? ',A<=B)  # меньше или равно
print (' A == B  ? ',A==B)  # равно
print (' A != B ? ',A!=B)   # не равно
print (' A < B < A+B ? ',A<B<A+B) # из интервала
print (' A < B и B < A+B  ? ',A<B and B<A+B) # и (*)
print (' A < B или A > B  ? ',A<B or A>B) # или (+)

print ('\n',A<B , A>B, A<B * A>B, A<B + A>B)  # \n - знак перехода на строку

print ('\n',bool(1), '\t', bool(0))  # \t - знак табуляции

 A >  B  ?  False
 A <  B  ?  True
 A >= B  ?  False
 A <= B  ?  True
 A == B  ?  False
 A != B ?  True
 A < B < A+B ?  True
 A < B и B < A+B  ?  True
 A < B или A > B  ?  True

 True False False True

 True 	 False


# 2 Основные инструкции

# 2.1 Операции ввода/вывода

Форматный вывод (это красиво): %ШиринаПоля.ТочностьТип

% - указатель на вставку, тип: i - int, f - float, s - str, e - "научное" представление, g - "компактное" представление десятичного числа

По умолчанию производится выравнивание по правому краю, для выравневания по левому нужно писать %- вместо %.

(Кроме того, существует метод str.format() для форматного вывода строк.)

In [21]:
A=1 
B=2
print(" A / B = %6.4f, возвращаемый тип - %s " % (A/B, type(A/B)))
print(" A / B = %-6i, возвращаемый тип - %s " % ((int)(A/B), type((int)(A/B))))

 A / B = 0.5000, возвращаемый тип - <class 'float'> 
 A / B = 0     , возвращаемый тип - <class 'int'> 


В интерактивной сессии (или в программе, запущенной с командной строки) можно попросить пользователя что-нибудь ввести. Аргумент функции input - это приглашение для ввода (prompt). Можно использовать просто input(), тогда приглашения не будет. Но это неудобно, т.к. в этом случае трудно заметить, что программа чего-то ждёт.

In [22]:
s=input('Введите число ')
print('s = ', s, type(s))
sf=float(s)
print('sf = ', sf, type(sf))

Введите число 4
s =  4 <class 'str'>
sf =  4.0 <class 'float'>


Питон - интерпретатор, поэтому он может во время выполнения программы интерпретировать строки как куски исходного текста на языке питон. Так, функция eval интерпретирует строку как выражение и вычисляет его (в текущем контексте - подставляя текущие значения переменных).

In [20]:
s=input('Введите операцию с переменной sf')
print('s = ', s, type(s))
print(s + ' = ', eval(s))

Введите операцию с переменной sfsf*2+1
s =  sf*2+1 <class 'str'>
sf*2+1 =  11.0


А функция exec интерпретирует строку как оператор и выполняет его. Оператор может менять значения переменных в текущем пространстве имён.

In [25]:
s=input('Введите выражение с переменной sf ')

Введите выражение с переменной sf x = sf - 1


In [27]:
s

'x = sf - 1'

In [28]:
exec(s)
print(x)

4.0


# 2.2 Ветвление (или условная инструкция)

In [ ]:
# Пример

a = input('Введите целое число a: ') 
b = input('Введите целое число b: ') 

if a < b: 
    print ('a - наименьшее из двух чисел') 
elif a == b: 
    print ('a и b равны')
else: 
    print ('b - наименьшее из двух чисел') 

In [ ]:
a = input('Введите целое число a: ') 
b = input('Введите целое число b: ') 

print ('a - наименьшее из двух чисел') if a<b else (print ('b - наименьшее из двух чисел') if a>b else print ('a и b равны'))

# 2.3 Цикл

# 2.4 Обработка ошибок

# 3 Функции

Простейшая в мире функция: не имеет параметров, ничего не делает и ничего не возвращает. Оператор pass означает "ничего не делай"; он используется там, где синтаксически необходим оператор, а делать ничено не нужно (после if или elif, после def и т.д.).

In [1]:
def fun0():
    pass

In [2]:
fun0()

# Ваша первая программа на Python

Традиционно первой программой при начале изучения того или иного языка программирования является программа, результатом работы которой является вывод на экран или иное устройство фразы «Hello, world!»

Её можно считать тестом для демонстрации работы системы перед переходом к более сложным вещам и объявлением сообществу (миру) о вашем появлении.

На разных языках программирования такая программа будет занимать разное колличество строк. В нашем случае достаточно написать:

print("Hello, world!")

Напишите здесь вашу первую программу на Python:

In [1]:
print("Hello, world!")

Hello, world!


# Пример 1. "Метание мяча"

Рассмотрим пример иллюстрирующий математическую модель описывающий траекторию полета мяча в воздухе. Из второго закона Ньютона, предполагая отсутствие сопротивления воздуха, мы можем вывести зависимость вертикального положения мяча от времени $t$:


$$y(t)=v_0t−\frac{1}{2}gt^2$$


где $v_0$ — начальная скорость, $g$ — ускорение свободного падения, значение которого положим равным 9.80665 $м/c^2$.

(Обратите внимание какие красивые формулы! Текстовые ячейки Jupiter поддерживают набор формул в синтаксесе вёрстки LaTeX, наиболее популярный набор макрорасширений (или макропакет) системы компьютерной вёрстки TeX.)

# Сценарий

Рассмотрим сценарий на Python для вычисления по этой простой формуле

In [14]:
# Программа для вычисления положения мяча при вертикальном движении

v0 = 5       # Начальная скорость
g = 9.80665  # Ускорение свободного падения
t = 0.6      # Время

y = v0*t - 1/2*g*t**2

print (y)

del v0, g, t, y

1.234803


# Разбор сценария или "как он это сделал?"

Сценарий на языке Python — это текст (или текстовый файл, в нашем случае эта программа написана и в файле ball.py, который можно запустить в редакторе Spyder), содержащий некоторые инструкции. Мы можем читать сценарий и понимать, что программа способна делать, но сам сценарий не выполняет никаких действий на компьютере, пока интерпретатор Python не прочитает текст сценария и не преобразует его в некоторые действия.

Когда сценарий запущен в интерпретаторе Python, он разбирается и выполняется построчно. То где вы это читаете не является 

Первая строчка:

In [91]:
# Программа для вычисления положения мяча при вертикальном движении

является комментарием, т.е. как только встречается символ #, интерпретатор Python воспринимает оставшуюся часть строки как комментарий, пропускает ее и переходит к следующей строке.

Следующая строка, интерпретируемая Pyhton:

In [92]:
v0 = 5       # Начальная скорость

В Python выражения вида v0 = 5 известны как оператции присваивания (с оператором "="). Значение правой части, в нашем случае целое число 5, становится объектом, а имя переменной слева — именованной ссылкой на этот объект. Всякий раз, когда мы запишем v0, Python заменит ее целым значением 5. Выражение v1 = v0 создает новое имя для того же целого объекта со значением 5, а не копирует объект. 

Следующие две строки

In [98]:
g = 9.80665  # Ускорение свободного падения
t = 0.6      # Время

имеют тот же смысл, что и предыдушая. Таким образом, прочитав эти строки интерпретатор Python знает три переменных v0, g, t и их значения. Эти переменные используются интерпретатором в следующей строке, реализующей некоторую формулу

In [99]:
y = v0*t - 1/2*g*t**2

В этой строке Python интерпретирует * как оператор умножения, - — вычитания, ** — возведения в степень (как и ^). Естественно, + и \ интерпретируются как операторы сложения и деления соответственно. В результате вычисляется значение по формуле в правой части выражения, которое присваивается переменной с именем y. 

Последняя строка сценария

In [100]:
print (y)

1.234803


выводит на печать значение переменной y. Таким образом при запуске сценария будет выведено число 1.234803. В тексте сценария имеются также и пустые строки, которые пропускаются интерпретатором Python. Они добавлены для лучшей читабельности кода.

# Упражнение 1. Сообщения об ошибках

Давайте сломаем работающую программу!!!


Предлагается ввести ошибки в сценарий. Для каждой введенной ошибки запустите сценарий. Напишите комментарий в тексте программы о том, как Python реагировал на соответствующую ошибку. После каждой ошибки поправьте сценарий так, чтобы он работал и переходите к следующей.

1. Вставьте слово Привет в пустой строке перед определением переменной v0.

2. Удалите знак # перед комментарием Начальная скорость

3. Удалите знак = в задании переменной v0

4. Замените слово print на pint

5. Замените строку print y на print x

6. Замените выражение g = 9.80665 и t = 0.1 на g: int = 9.80665 и t: int = 0.1, а потом на g = int(9.80665) и t = int(0.1)


Проверьте Python на прочность:

In [15]:
# Программа для вычисления положения мяча при вертикальном движении

v0 = 5       # Начальная скорость (м/с)
g = 9.80665  # Ускорение свободного падения (м/с^2)
t = 0.6      # Время (с)

y = v0*t - 1/2*g*t**2 # (м)

print (y)

del v0, g, t, y

1.234803


In [16]:
g = 9.80665 
g

9.80665

In [17]:
g: int = 9.80665
g

9.80665

In [18]:
g = int(9.80665)
g

9

# Упражнение 2. Форматный вывод результата

In [ ]:
# Программа для вычисления положения мяча при вертикальном движении

v0 = 5       # Начальная скорость
g = 9.80665  # Ускорение свободного падения
t = 0.6      # Время

y = v0*t - 1/2*g*t**2

print (y)

del v0, g, t, y

# Упражнение 3. Интерактивное общение с пользователем

Реализовать в программе запрос исходных данных (хотя бы одной) с помощью интерактивной сессии функцией input().

In [5]:
data = input()

# Упражнение 4. Проверка корректности исходных данных

Заметим, что в рассматриваемой задаче ось $y$ выбрана так, что мячик в начальный момент времени ($t = 0$) находится в точке с координатой $y = 0$. Чтобы посмотреть сколько времени займет у мячика вернуться в эту же координату ($y = 0$), надо рассмотреть следующее вырожение:

$$v_0t−\frac{1}{2}gt^2 = t \left( v_0t−\frac{1}{2}gt \right) = 0$$

Последнее возможно только если $$t = 0$$ или $$t = \frac{2v_0}{g}.$$

Таким образом возможное время полета мяча ограничено интервалом

$$0 \leqslant t \leqslant \frac{2v_0}{g}.$$

Добавить в программу проверку корректности вводимых пользователем переменных, исходя из условий: начальная скорость и ускорение свободного падения не отрицальны, диапазон возможного времени определяется полученным выше условием.

In [ ]:
if v0>0 and g>0 and t>=0 and t<=(2*v0/g):
    print("Данные корректны")
else:
    print("ОШИБКА:Некорректные данные!")

# Домашнее задание (базовое):

# Задание 1. Обновление переменных в программной оболочке Python

Написать программу реализующую следующие действия:

1. Инициализировать переменную x значением 2
2. Прибавить 3 к x и присвоить это значение x
3. Вычислить (x+1)/2 и присвоить это значение x
4. Вычислить x+1/2 и присвоить это значение x 
5. Присвоить x значение значение выражения x<5 
6. Изменить тип x в символьный и присвоить результат переменной x

После каждой операции выводить значение переменной x и её тип.

In [9]:
x = 2
print(x,type(x))
x+=3
print(x,type(x))
x = (x+1)/2
print(x,type(x))
x = x+1/2
print(x,type(x))
x = x<5
print(x,type(x))
x = str(x)
print(x,type(x))

2 <class 'int'>
5 <class 'int'>
3.0 <class 'float'>
3.5 <class 'float'>
True <class 'bool'>
True <class 'str'>


# Задание 2. Среднее значение целых чисел

Написать программу (сценарий), сохраняющую сумму 1+2+3+4+5 в одну переменную, затем создающую другую переменную, содержащую среднее значение этих пяти чисел. Вывести фразу "Среднее значение этих 5 чисел равно" и результат работы с точностью до 5 знака после точки (т.е. до 1/100000).

In [15]:
n = 1+2+3+4+5
a = n/5
print("Среднее значение этих 5 чисел равно",a)



Среднее значение этих 5 чисел равно 3.0


# Задание 3. Интерактивное вычисление среднего

Добавьте в программу из предыдущего задания запрос на добавление ещё однонго числа к подсчёту среднего с помощью запуска интерактивной сессии функцией input(). После введения каждого нового числа пересчитывать и выводить резёльтата в формате: "Среднее значение этих (число чисел) чисел равно" и результат работы с точностью до 5 знака после точки (т.е. до 1/100000). Интерактивная сессия заканчивается после введения пользователем 0.

In [17]:
n = 1+2+3+4+5
b = int(input())
if b!=0:
    n +=b
    a = n/6
    print("Среднее значениие этих 6 чисел равно",a)



Среднее значениие этих 6 чисел равно 2.6666666666666665


# Домашнее задание (дополнительное):

# Задание. "О чём думает кот?"

In [20]:
# Так можно добавлять картинки

from IPython.display import Image # вызов из библиотеки определённой функции
Image("picture.png")              # вызов функции и передача ей в качестве аргумента пути к файлу 

# (в данном случае фаил находится в той же папке)

FileNotFoundError: No such file or directory: 'picture.png'

FileNotFoundError: No such file or directory: 'picture.png'

<IPython.core.display.Image object>

Рассмотрим пример иллюстрирующий математическую модель описывающий траекторию кошачьего прыжка. Из второго закона Ньютона, предполагая отсутствие сопротивления воздуха, мы можем вывести зависимость вертикального (y) и горизонтального (x) положения котика от времени $t$:

$$x(t)=v_0cos(\alpha)t$$
$$y(t)=v_0sin(\alpha)t−\frac{1}{2}gt^2$$

где $v_0$ — начальная скорость, $g$ — ускорение свободного падения, значение которого положим равным 9.80665 $м/c^2$, $tg(\alpha) = y/x$.

Напишите программу которая сможет помочь котику понять следующие вещи:


1) как далеко и высоко он сможет прыгнуть при начальной скорости 5 м/с и направлении пражка $\alpha = 40^{\circ}$;

2) каким должен быть начальный угол для максимальной дальности прыжка;

3) каким должен быть начальный угол для максимальной высоты прыжка;

4) какимо должны быть угол и скорость для попадания в точку с координатами (2,3) м через 1 с;

5) нарисуйте траекторию прыжка с помощью библиотеки matplotlib.


# Задание. "Число и цифра"

Запросите у пользователя целое число и сложите цифры этого числа. Повторяйте последнее действие столоко раз сколько сможете и выводите каждый раз полученную сумму.

In [18]:
while True:
    n = int(input())
    a = 0
    while n>0:
        a+=n%10
        n = n//10
    print(a)

3
5
7
10


KeyboardInterrupt: Interrupted by user

# Задание. "Максимум и минимум"

Запрашивайте у пользователя числа, определяйте максимальное и минимальной из всех полученных чисел и сообщайте их ему.

In [19]:
a = int(input())
b = int(input())
c = int(input())
d = int(input())
e = int(input())
print("Максимальное число:",max(a,b,c,d,e))
print("Минимльное число:",min(a,b,c,d,e))

Максимальное число: 43314
Минимльное число: -100
